homeassistant/core.py

```diff
@@ -1907,18 +1932,27 @@ async def async_call(
             raise asyncio.CancelledError
         if task.done():
             # Propagate any exceptions that might have happened during service call.
-            task.result()
+            result = task.result()
             # Service call completed successfully!
+            if return_values:
+                if not isinstance(result, dict):
+                    raise HomeAssistantError(
+                        f"Service return value expected a dictionary, was {type(result)}"
+                    )
+                return result
             return True
+
         # Service call task did not complete before timeout expired.
         # Let it keep running in background.
         self._run_service_in_background(task, service_call)
         _LOGGER.debug("Service did not complete before timeout: %s", service_call)
-        return False
+        raise asyncio.TimeoutError(
+            f"Service did not complete before timeout: {repr(service_call)}",
+        )
 
     def _run_service_in_background(
         self,
-        coro_or_task: Coroutine[Any, Any, None] | asyncio.Task[None],
+        coro_or_task: Coroutine[Any, Any, ServiceResult] | asyncio.Task[ServiceResult],
```

Comment:
Running in the background should not support a service return result, right?

Reply:
I get the point and agree in theory. This function does not return anything. It accepts a coroutine or task that does because thats whats services do., but concretely: Is the proposal that this functions signature change should be reverted to `None` and the caller should cast or ignore the `arg-type` warning? or something else?

```
homeassistant/core.py:1915: error: Argument 1 to "_run_service_in_background" of "ServiceRegistry" has incompatible type "Coroutine[Any, Any, Union[bool, Dict[str, Any], None]]"; expected "Union[Coroutine[Any, Any, None], Task[None]]"  [arg-type]
homeassistant/core.py:1947: error: Argument 1 to "_run_service_in_background" of "ServiceRegistry" has incompatible type "Task[Union[bool, Dict[str, Any], None]]"; expected "Union[Coroutine[Any, Any, None], Task[None]]"  [arg-type]
```